neqo-transport/src/connection/idle.rs

```diff
@@ -96,6 +96,21 @@ impl IdleTimeout {
         self.start(now) + max(self.timeout / 2, pto)
     }
 
+    pub fn maybe_keep_alive_timeout(&self, now: Instant, pto: Duration) -> Option<Instant> {
+        if self.keep_alive_outstanding {
+            return None;
+        }
+
+        let timeout = self.keep_alive_timeout(now, pto);
+        // This could happen when the state is AckElicitingPacketSent(t) and
+        // the t exceeds keep_alive_timeout.
+        if timeout <= now {
+            return None;
+        }
```

Comment:
Good point!
The code for the case you described can be found here:
https://github.com/mozilla/neqo/blob/main/neqo-transport/src/connection/idle.rs#L105-L107
I do agree that we should send a keep-alive for this case.

However, when this function `maybe_keep_alive_timeout` is called, `maybe_probe` was already called before and we have nothing to send. At this point, we need to determine the appropriate timeout value.
I think it's fine to:
1. return `None`
2. return a timeout that is smaller than idle timeout. Maybe `now + max(self.timeout / 2, pto)`?
What do you think?

Reply:
I have some pending changes to this in #2129:
https://github.com/mozilla/neqo/blob/69bb7f89fe55ab3e4153be3f4a007887a301ec03/neqo-transport/src/connection/mod.rs#L2121-L2164

Reply:
In what case would the line above execute, in other words, in what case would we have no outstanding keep-alive, and an ack-eliciting packet more than keep-alive-timeout ago, but not send a keep-alive in `Connection::maybe_probe`?

In other words what happens if we delete the conditional?
```suggestion
```

If there is no valid scenario, we should add an `assert!` instead of the `return None`, or simply return `timeout` and depend on the `debug_assert` in `Connection`:

https://github.com/mozilla/neqo/blob/55e3a9363c28632dfb29ce91c7712cab1f6a58da/neqo-transport/src/connection/mod.rs#L1098

Reply:
This case can be triggered by running this test: `connection::tests::idle::keep_alive_with_unresponsive_server`
We will hit the debug assert `debug_assert!(earliest > now);` after removing the code.

Looking at the code in `maybe_probe`, we don't send a PING because the function returns early here:
https://github.com/mozilla/neqo/blob/69bb7f89fe55ab3e4153be3f4a007887a301ec03/neqo-transport/src/connection/mod.rs#L2133-L2137

I am not sure if this is a bug, since the comment seems to suggest that we should send a PING, but actually we don't.

Reply:
Thanks for the reference to the test.

Debugging [`keep_alive_with_unresponsive_server`](https://github.com/mozilla/neqo/blob/55e3a9363c28632dfb29ce91c7712cab1f6a58da/neqo-transport/src/connection/tests/idle.rs#L737-L757) further it seems like:

1. In the last iteration, i.e. where we hit `timeout <= now`, we are cc limited, thus not writing any ackeliciting packets or probe packets even though the idle timeout is due.
    https://github.com/mozilla/neqo/blob/55e3a9363c28632dfb29ce91c7712cab1f6a58da/neqo-transport/src/connection/mod.rs#L2212-L2215
2. After failing to send anything, we call `Connection::next_delay` and the above `IdleTimeout::maybe_keep_alive_timeout`. Given that the idle timeout is still due, we hit `timeout <= now`.

If I understand [RFC 9000 12.4](https://www.rfc-editor.org/rfc/rfc9000.html#section-12.4) correctly, the above (1) is correct behavior, i.e. a `PING` frame does count towards bytes in-flight and thus has to adhere to congestion control.

With the above in mind, I am in favor of the original change proposed in this pull request, namely to return `None`, i.e. no keep alive timeout. I just suggest updating the comment.

```suggestion
        // Timer is in the past, i.e. we should have sent a keep alive, but we were unable to, e.g. due to CC.
        if timeout <= now {
            return None;
        }
```

@KershawChang @larseggert thoughts?

Reply:
Sounds good to me. Will update the PR later.

Reply:
I think we can **always** send one or two packets on PTO, even if we are CC limited. So we should be able to probe?

Reply:
Why would we need to send a keepalive if we were limited by the congestion controller?  That would imply that we were sending a bunch of stuff anyway.  I believe that the lowest send rate we could achieve is 2 packets per RTT, which is much higher than one at half the idle timeout.